File: TouchNothing/Services/LeaderboardService.swift
import Foundation

enum LeaderboardLoadStatus: Equatable {
    case success
    case offlineCached
    case failed
}

struct LeaderboardLoadResult {
    let rows: [LeaderboardRow]
    let status: LeaderboardLoadStatus

    var statusMessage: String? {
        switch status {
        case .success, .failed:
            return nil
        case .offlineCached:
            return LocalizationKey.leaderboardOffline.localized
        }
    }

    var isFailure: Bool {
        status == .failed
    }
}

enum LeaderboardService {
    static func load(
        mode: LeaderboardMode,
        currentNickname: String?
    ) async -> LeaderboardLoadResult {
        switch mode {
        case .bestSession:
            await loadBestSession(currentNickname: currentNickname)
        case .totalTime:
            await loadTotal(currentNickname: currentNickname)
        }
    }

    private static func loadBestSession(currentNickname: String?) async -> LeaderboardLoadResult {
        do {
            let entries = try await SupabaseService.shared.fetchLeaderboard()
            LeaderboardCache.saveBestSession(entries: entries)
            return LeaderboardLoadResult(
                rows: LeaderboardRowMapper.map(entries: entries, currentNickname: currentNickname),
                status: .success
            )
        } catch {
            return cachedResult(
                cachedEntries: LeaderboardCache.loadBestSession(),
                currentNickname: currentNickname,
                mapRows: LeaderboardRowMapper.map(entries:currentNickname:)
            )
        }
    }

    private static func loadTotal(currentNickname: String?) async -> LeaderboardLoadResult {
        do {
            let entries = try await SupabaseService.shared.fetchTotalLeaderboard()
            LeaderboardCache.saveTotal(entries: entries)
            return LeaderboardLoadResult(
                rows: LeaderboardRowMapper.mapTotal(entries: entries, currentNickname: currentNickname),
                status: .success
            )
        } catch {
            return cachedResult(
                cachedEntries: LeaderboardCache.loadTotal(),
                currentNickname: currentNickname,
                mapRows: LeaderboardRowMapper.mapTotal(entries:currentNickname:)
            )
        }
    }

    private static func cachedResult<T>(
        cachedEntries: [T],
        currentNickname: String?,
        mapRows: ([T], String?) -> [LeaderboardRow]
    ) -> LeaderboardLoadResult {
        guard !cachedEntries.isEmpty else {
            return LeaderboardLoadResult(rows: [], status: .failed)
        }

        return LeaderboardLoadResult(
            rows: mapRows(cachedEntries, currentNickname),
            status: .offlineCached
        )
    }
}

File: app/src/main/java/com/touchnothing/app/domain/LeaderboardService.kt
package com.touchnothing.app.domain

import com.touchnothing.app.data.local.LeaderboardCache
import com.touchnothing.app.data.model.LeaderboardMode
import com.touchnothing.app.data.model.LeaderboardRow
import com.touchnothing.app.data.remote.SupabaseService

class LeaderboardService(
    private val leaderboardCache: LeaderboardCache,
    private val supabaseService: SupabaseService = SupabaseService.shared,
) {
    suspend fun load(mode: LeaderboardMode, currentNickname: String?): LeaderboardLoadResult =
        when (mode) {
            LeaderboardMode.BEST_SESSION -> loadBestSession(currentNickname)
            LeaderboardMode.TOTAL_TIME -> loadTotal(currentNickname)
        }

    private suspend fun loadBestSession(currentNickname: String?): LeaderboardLoadResult {
        return runCatching {
            val entries = supabaseService.fetchLeaderboard()
            leaderboardCache.saveBestSession(entries)
            LeaderboardLoadResult(
                rows = LeaderboardRowMapper.map(entries, currentNickname),
                status = LeaderboardLoadStatus.SUCCESS,
            )
        }.getOrElse {
            cachedResult(
                cachedEntries = leaderboardCache.loadBestSession(),
                currentNickname = currentNickname,
            ) { entries, nickname ->
                LeaderboardRowMapper.map(entries, nickname)
            }
        }
    }

    private suspend fun loadTotal(currentNickname: String?): LeaderboardLoadResult {
        return runCatching {
            val entries = supabaseService.fetchTotalLeaderboard()
            leaderboardCache.saveTotal(entries)
            LeaderboardLoadResult(
                rows = LeaderboardRowMapper.mapTotal(entries, currentNickname),
                status = LeaderboardLoadStatus.SUCCESS,
            )
        }.getOrElse {
            cachedResult(
                cachedEntries = leaderboardCache.loadTotal(),
                currentNickname = currentNickname,
            ) { entries, nickname ->
                LeaderboardRowMapper.mapTotal(entries, nickname)
            }
        }
    }

    private fun <T> cachedResult(
        cachedEntries: List<T>,
        currentNickname: String?,
        mapRows: (List<T>, String?) -> List<LeaderboardRow>,
    ): LeaderboardLoadResult {
        if (cachedEntries.isEmpty()) {
            return LeaderboardLoadResult(emptyList(), LeaderboardLoadStatus.FAILED)
        }
        return LeaderboardLoadResult(
            rows = mapRows(cachedEntries, currentNickname),
            status = LeaderboardLoadStatus.OFFLINE_CACHED,
        )
    }
}
